src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -261,43 +251,88 @@ static DiscoveryCriteria NewDiscoveryCriteriaFromSourceAndSettings(IEnumerable<s
                 proxyDiscoveryManager.DiscoverTests(discoveryCriteria, eventHandler);
-            })
-            .ContinueWith(t =>
-                {
-                    // Just in case, the actual discovery couldn't start for an instance. Ensure that
-                    // we call discovery complete since we have already fetched a source. Otherwise
-                    // discovery will not terminate
-                    EqtTrace.Error("ParallelProxyDiscoveryManager: Failed to trigger discovery. Exception: " + t.Exception);
-
-                    var handler = eventHandler;
-                    var exceptionToString = t.Exception?.ToString();
-                    var testMessagePayload = new TestMessagePayload { MessageLevel = TestMessageLevel.Error, Message = exceptionToString };
-                    handler.HandleRawMessage(_dataSerializer.SerializePayload(MessageType.TestMessage, testMessagePayload));
-                    handler.HandleLogMessage(TestMessageLevel.Error, exceptionToString);
-
-                    // Send discovery complete. Similar logic is also used in ProxyDiscoveryManager.DiscoverTests.
-                    // Differences:
-                    // Total tests must be zero here since parallel discovery events handler adds the count
-                    // Keep `lastChunk` as null since we don't want a message back to the IDE (discovery didn't even begin)
-                    // Set `isAborted` as true since we want this instance of discovery manager to be replaced
-                    // TODO: the comment above mentions 0 tests but sends -1. Make sense of this.
-                    var discoveryCompleteEventsArgs = new DiscoveryCompleteEventArgs(-1, true);
-                    handler.HandleDiscoveryComplete(discoveryCompleteEventsArgs, null);
-                },
-                TaskContinuationOptions.OnlyOnFaulted);
+            }).ContinueWith(t => HandleError(eventHandler, t), TaskContinuationOptions.OnlyOnFaulted);
 
         EqtTrace.Verbose("ProxyParallelDiscoveryManager.DiscoverTestsOnConcurrentManager: No sources available for discovery.");
     }
 
+    private void HandleError(ITestDiscoveryEventsHandler2 eventHandler, Task t)
+    {
+        // Just in case, the actual discovery couldn't start for an instance. Ensure that
+        // we call discovery complete since we have already fetched a source. Otherwise
+        // discovery will not terminate
+        EqtTrace.Error("ParallelProxyDiscoveryManager: Failed to trigger discovery. Exception: " + t.Exception);
+
+        var handler = eventHandler;
+        var testMessagePayload = new TestMessagePayload { MessageLevel = TestMessageLevel.Error, Message = t.Exception?.ToString() };
+        handler.HandleRawMessage(_dataSerializer.SerializePayload(MessageType.TestMessage, testMessagePayload));
+        handler.HandleLogMessage(TestMessageLevel.Error, t.Exception?.ToString());
```

Comment:
Avoid double call to `ToString`.

```suggestion
        var exceptionString = t.Exception?.ToString();
        var testMessagePayload = new TestMessagePayload { MessageLevel = TestMessageLevel.Error, Message = exceptionString };
        handler.HandleRawMessage(_dataSerializer.SerializePayload(MessageType.TestMessage, testMessagePayload));
        handler.HandleLogMessage(TestMessageLevel.Error, exceptionString);
```